Primitives for collections

dvisztempacct opened the pull request:
I'd like to see Future primitives for different sorts of collections. This is my attempt at supporting `Bluebird.map` sort of interface for arrays. Feedback greatly appreciated!

dvisztempacct committed:
added flatMapArray

dvisztempacct committed:
ignore unused iterator binding

dvisztempacct commented on src/FutureResult.re:
```diff
@@ -1,5 +1,5 @@
 let deprecate = (ft, name, more) =>
-  ft |. Future.tap(_ => Js.Console.warn("FutureResult."++name++" is deprecated." ++ more))
+  ft |. Future.tap(_ => Js.Console.warn("FutureResult."++name++" is deprecated." ++ more));
```
whoops.... i didn't mean to commit this. pretty sure i accidentally `bsrefmt`d it

dvisztempacct committed:
ignore unused iterator binding

dvisztempacct committed:
undo refmt

scull7 reviewed (changes requested):
Would you please add some tests for this?

dvisztempacct commented:
More downsides to my approach: requires `Belt` and `Obj.magic` are those dependencies we want `Future` to have?

gilbert commented:
Future already supports `Belt`, so that's ok, but `Obj.magic` is probably a dep we don't want.

dvisztempacct committed:
Merge branch 'master' of github.com:RationalJS/future

dvisztempacct commented:
I filed issue #16 before remembering this PR...

any possibility this will get merged or something similar get added?

dvisztempacct committed:
addressed type-safety concerns in flatMap

dvisztempacct commented:
I've renamed `flatMapArray` to `flatMapArrayUnsafe` because of its use of `Obj.magic` and `Belt.Array.makeUninitialized`.

I've added a new `flatMapArray` that uses an `array(option('a))` first, and then maps that array back to an `array('t)` to resolve the future.

To see how to use these functions, see my example code:

https://github.com/RationalJS/future/issues/16#issuecomment-397950948

dvisztempacct committed:
added "all" functions

dvisztempacct commented:
Oh, it occurs to me that I haven't really solved the problem if the problem is an added dependency, as opposed to use of unsafe functions.

@gilbert let me know!

@scull7 the existing tests are broken. I think they assume too much about how bucklescript represents certain values in Javascript and how those get converted to strings. Simply adding my functions and adding no tests for my functions breaks the existing tests. @gilbert what do you think?

leostera commented on src/Future.re:
```diff
@@ -75,3 +75,74 @@ let tapError = (future, f) => future |. tap(r => switch(r) {
   | Ok(_) => ()
 });
+
+let flatMapArray = (xs: array('a), f: 'a => t('b), maxConcurrent) =>
```
This feels like tying a few things together: batching the future creation, executing the batches one after another, and collecting the results.

I'd have expected something like:

```reason

/* this is the part where I prepare my data for the concurrent execution */
module type List = {
  /* take a list of 'a, and an int for how many to put in each bucket */
  let chunks_of: (int, list('a)) => list(list('a));
};

let futures = [1, 2, 3, 4] |. List.map(x => () => Future.value(x)); /* this could also be lazy futures */
let maxConcurrency = 2;
let chunks: list(list(unit => Future.t(int)) = List.chunks_of(maxConcurrency, futures);
/* [ [() => Future(1), () => Future(2)], [() => Future(3), () => Future(4)] ] */

/* this is the part where I build up the computation */
module type Future = {
  let all: (list(t('a)), list('a) => 'b); /* from https://github.com/RationalJS/future/issues/16 */

  let sequence: list(list(unit => Future.t('a))) => Future.t(list(list('a)) = chunks => {
    let rec consumer = (cs, acc) => {
      switch (cs) {
      | [] => Future.value(acc);
      | [h, ...t] => {
          /* initialize all futures in this chunk: this would be more elegant using Lazy/Force */
         Future.all(List.map(x => x(), h))
         |. Future.flatMap(results => consumer(t, [results, ...acc])))
        }
      }
    };
    consumer(chunks, []);
    };
};

/* this is the part where you consume it */
Future.sequence(chunks).get( values => Js.log(values) );
```

Of course this is just a sketch but it's a bit more elegant, it (from what I can see) won't throw any exceptions anywhere, it can still be optimized with mutable cursors and good ol' `while` loops if the performance needs it.

Just a thought! Good work @dvisztempacct on pushing for concurrency idioms in the library 🙌

dvisztempacct replied:
Thanks for the feedback @ostera 

I think your proposal has a limiting factor, however: each "chunk" must be processed serially. I know from practical experience that this can seriously limit performance.

dvisztempacct replied:
This sort of design also gives the allocator a much greater workout. My original implementation (now called `flatMapArrayUnsafe`) only uses a single array of length `n` and (I believe) a single closure scope and two functions with access to that scope.

@ostera do you think you could provide a full implementation of your proposal so I could more closely study its performance characteristics?

Thanks :)

leostera replied:
> @dvisztempacct: each "chunk" must be processed serially

Oh, perhaps I misinterpreted the idea behind your `maxConcurrent`. I thought it to mean that Future's should be batched in groups of at most `maxConcurrent`.

Naturally if that wasn't the goal the implementation above can be changed by replacing the `Future.sequence` call with a `Scheduler`/`TaskRunner` that can take care of this concern.

Thanks for pointing this out! 💯 

> @dvisztempacct: This sort of design also gives the allocator a much greater workout

If by allocator you mean whoever uses this library, yes. If you mean the runtime, you'd still be creating a function for every `xs[i] |. f |. get( x => ... )` 🤔. Optimizations on both ends can be made to reuse the functions (in your case you have everything already in the upper scope, so it should be trivial to extract). It'd be interesting to compare the compiled Javascript of both solutions.

In the end your implementation will run faster than mine by the nature of it's design. I'll extract a Scheduler from it and perhaps then we can compare more fairly 👍

dvisztempacct replied:
Thanks @ostera

scull7 replied:
@ostera https://github.com/jonlaing/rationale/blob/master/src/Monad.re. Is this similar to what you are looking for?

leostera replied:
@scull7 essentially this bit -> https://github.com/jonlaing/rationale/blob/master/src/Monad.re#L48-L54

leostera replied:
But it's different from what @dvisztempacct implemented tho! His Future pool keeps a constant number of promises running, whereas using `all` or `all_ignore` will run them in sequence, one by one AFAICT.

gilbert commented:
> Simply adding my functions and adding no tests for my functions breaks the existing tests.

Did #18 fix this issue?

dvisztempacct commented:
>Did #18 fix this issue?

@gilbert yes

dvisztempacct commented:
Closing. Was a similar facility ever added?
